python/src/equistore/operations/unique_metadata.py

```diff
@@ -0,0 +1,245 @@
+    :param names: a str, list of str, or tuple of str corresponding to the
+        name(s) of the metadata along the specified ``axis`` for which the
+        unique indices should be found.
+    :param gradient_param: a str corresponding to the gradient parameter name
+        for the gradient blocks to find the unique metadata for. If none
+        (default), the unique metadata of the regular TensorBlocks will be
+        calculated.
+
+    :return: a sorted :py:class:`Labels` object containing the unique metadata
+        for the input ``block`` or its gradient for the specified parameter.
+        Each element in the returned :py:class:`Labels` object has len(names)
+        entries.
+    """
+    # Parse input args
+    if not isinstance(block, TensorBlock):
+        raise TypeError("``block`` must be an equistore TensorBlock")
+    names = (
+        [names]
+        if isinstance(names, str)
+        else (list(names) if isinstance(names, tuple) else names)
+    )
+    _check_args(block, axis, names, gradient_param)
+    # Make a list of the blocks to find unique indices for
+    if gradient_param is None:
+        blocks = [block]
+    else:
+        blocks = [block.gradient(gradient_param)]
+
+    return _unique_from_blocks(blocks, axis, names, gradient_param)
+
+
+def _unique_from_blocks(
+    blocks: List[TensorBlock],
+    axis: str,
+    names: List[str],
+    gradient_param: Optional[str],
+) -> Labels:
+    """
+    Finds the unique metadata of a list of blocks along the given ``axis`` and
+    for the specified ``names``. If ``gradient_param`` is specified, only finds
+    the unique indices for gradient blocks under the specified parameter name.
+    """
+    # Extract indices from each block
+    all_idxs = []
+    for block in blocks:
+        idxs = block.samples[names] if axis == "samples" else block.properties[names]
+        for idx in idxs:
+            all_idxs.append(idx)
+
+    # If no matching indices across all blocks return a empty Labels w/ the
+    # correct names
+    if len(all_idxs) == 0:
+        # Create Labels with single entry
+        labels = Labels(names=names, values=np.array([[i for i in range(len(names))]]))
+        # rslice to zero length
+        return labels[:0]
+
+    # Define the unique and sorted indices
+    unique_idxs = np.unique(all_idxs, axis=0)
+
+    # Return as Labels
+    return Labels(names=names, values=np.array([[j for j in i] for i in unique_idxs]))
+
+
+def _check_args(
+    tensor: Union[TensorMap, TensorBlock],
+    axis: str,
+    names: List[str],
+    gradient_param: Optional[str] = None,
+):
+    """Checks input args for :py:func:`unique_metadata` and
+    :py:func:`unique_metadata_block`."""
+    # Check tensors
+    if isinstance(tensor, TensorMap):
+        blocks = tensor.blocks()
+        # Check gradients
+        if gradient_param is not None:
+            if not isinstance(gradient_param, str):
+                raise TypeError("``gradient_param`` must be a str")
```

Comment:
since this will not be rendered with rst we can skip the double backticks  \`\`
```suggestion
                raise TypeError("`gradient_param` must be a str")
```